### ISSUE TITLE
Migration: Changing an index that is part of an FK relationship fails

### ISSUE THREAD

**Issue by LordMike** [PomeloFoundation/Pomelo.EntityFrameworkCore.MySql#678]
### Steps to reproduce

1. Create two types:
  * A: int Id, int BId; use an AK on A.BId
  * B: int Id;
2. Create an FK relationsip. A.BId -> B.Id
3. Create Migration
4. Alter A so that it no longer uses an AK but rather a Unique Index (I was troubleshooting the exact same issue as in #90)
5. Create migration

### The issue
In the final result, Pomelo should have generated a "DROP CONSTRAINT" statement (which is wrong - but nevermind that), without first removing the FK constraint. The DROP will fail, with mysql stating that it can't be done.

I've workarounded it by dropping the FK, removing the AK, adding a unique index, then re-creating the FK.

Attached here, [ConstraintsIssue.zip](https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/files/2346374/ConstraintsIssue.zip), is a solution with the issue.

Using `dotnet ef migrations script`, it will create the following:

```mysql
CREATE TABLE `__EFMigrationsHistory` (
    `MigrationId` VARCHAR(95) NOT NULL,
    `ProductVersion` VARCHAR(32) NOT NULL,
    CONSTRAINT `PK___EFMigrationsHistory` PRIMARY KEY (`MigrationId`)
);

CREATE TABLE `Bs` (
    `Id` INT NOT NULL AUTO_INCREMENT,
    CONSTRAINT `PK_Bs` PRIMARY KEY (`Id`)
);

CREATE TABLE `As` (
    `Id` INT NOT NULL AUTO_INCREMENT,
    `BId` INT NOT NULL,
    CONSTRAINT `PK_As` PRIMARY KEY (`Id`),
    CONSTRAINT `AK_As_BId` UNIQUE (`BId`),
    CONSTRAINT `FK_As_Bs_BId` FOREIGN KEY (`BId`) REFERENCES `Bs` (`Id`) ON DELETE CASCADE
);

INSERT INTO `__EFMigrationsHistory` (`MigrationId`, `ProductVersion`)
VALUES ('20180903184926_Initial', '2.1.2-rtm-30932');

-- ALTER TABLE `As` DROP CONSTRAINT `AK_As_BId`; -- Ignore this error for now
ALTER TABLE `As` DROP INDEX `AK_As_BId`;

CREATE UNIQUE INDEX `IX_As_BId` ON `As` (`BId`);

INSERT INTO `__EFMigrationsHistory` (`MigrationId`, `ProductVersion`)
VALUES ('20180903184954_ChangeIndexType', '2.1.2-rtm-30932');
```

In this, `ALTER TABLE As DROP INDEX AK_As_BId;` will fail.

### Further technical details

MySQL version: `5.7.23`
Operating system: Windows 10
Pomelo.EntityFrameworkCore.MySql version: `2.1.1`

**Comment by ruant** [comment]
Bumped into this issue too.
Changed some indexes from being unique to non-unique, but since they all are in foreign key constraints they won't update since the FK's aren't dropped before the index is dropped and recreated.

I guess hand editing the migrations would be the quickest way to get stuff moving along for now (https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/issues/591#issuecomment-400871313)

**Comment by FWest98** [comment]
This problem also occurs when changing (dropping & recreating) the primary key of a table. The combination of primary key being also a FK constraint is not very common, but might occur more frequently with the Owned Types in EF Core.

**Comment by PawelGerr** [comment]
Same issue using MariaDb 10.4.7

* Pomelo.EntityFrameworkCore.MySql: 3.0.0-rc1.final

### Migration
```
migrationBuilder.DropPrimaryKey("PK_MyTable", "MyTable");
```

`MyTable` has a composite PK: `OtherTypeId, UserId`. `OtherTypeId` is FK to the table `OtherTable`.

### Exception
```
MySql.Data.MySqlClient.MySqlException (0x80004005): Error on rename of '.\myproject\#sql-1a64_1d' to '.\myproject\mytable' (errno: 150 "Foreign key constraint is incorrectly formed")
 ---> MySql.Data.MySqlClient.MySqlException (0x80004005): Error on rename of '.\myproject\#sql-1a64_1d' to '.\myproject\mytable' (errno: 150 "Foreign key constraint is incorrectly formed")
   at MySqlConnector.Core.ResultSet.ReadResultSetHeaderAsync(IOBehavior ioBehavior) in C:\projects\mysqlconnector\src\MySqlConnector\Core\ResultSet.cs:line 49
   at MySql.Data.MySqlClient.MySqlDataReader.ActivateResultSet() in C:\projects\mysqlconnector\src\MySqlConnector\MySql.Data.MySqlClient\MySqlDataReader.cs:line 125
   at MySql.Data.MySqlClient.MySqlDataReader.CreateAsync(CommandListPosition commandListPosition, ICommandPayloadCreator payloadCreator, IDictionary`2 cachedProcedures, IMySqlCommand command, CommandBehavior behavior, IOBehavior ioBehavior, CancellationToken cancellationToken) in C:\projects\mysqlconnector\src\MySqlConnector\MySql.Data.MySqlClient\MySqlDataReader.cs:line 338
   at MySqlConnector.Core.CommandExecutor.ExecuteReaderAsync(IReadOnlyList`1 commands, ICommandPayloadCreator payloadCreator, CommandBehavior behavior, IOBehavior ioBehavior, CancellationToken cancellationToken) in C:\projects\mysqlconnector\src\MySqlConnector\Core\CommandExecutor.cs:line 63
   at MySql.Data.MySqlClient.MySqlCommand.ExecuteNonQueryAsync(IOBehavior ioBehavior, CancellationToken cancellationToken) in C:\projects\mysqlconnector\src\MySqlConnector\MySql.Data.MySqlClient\MySqlCommand.cs:line 227
   at Microsoft.EntityFrameworkCore.Storage.RelationalCommand.ExecuteNonQueryAsync(RelationalCommandParameterObject parameterObject, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Storage.RelationalCommand.ExecuteNonQueryAsync(RelationalCommandParameterObject parameterObject, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Storage.RelationalCommand.ExecuteNonQueryAsync(RelationalCommandParameterObject parameterObject, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Migrations.Internal.MigrationCommandExecutor.ExecuteNonQueryAsync(IEnumerable`1 migrationCommands, IRelationalConnection connection, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Migrations.Internal.MigrationCommandExecutor.ExecuteNonQueryAsync(IEnumerable`1 migrationCommands, IRelationalConnection connection, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Migrations.Internal.MigrationCommandExecutor.ExecuteNonQueryAsync(IEnumerable`1 migrationCommands, IRelationalConnection connection, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Migrations.Internal.MigrationCommandExecutor.ExecuteNonQueryAsync(IEnumerable`1 migrationCommands, IRelationalConnection connection, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Migrations.Internal.Migrator.MigrateAsync(String targetMigration, CancellationToken cancellationToken)
   at myproject.DbContextSetup.MigrateAsync(IServiceProvider provider) in C:\Projects\myproject\DbContextSetup.cs:line 64
```
### Logs
```
[15:19:28 INF] [Microsoft.EntityFrameworkCore.Database.Command] Executing DbCommand [Parameters=[], CommandType='Text', CommandTimeout='30']
DROP PROCEDURE IF EXISTS POMELO_BEFORE_DROP_PRIMARY_KEY;
CREATE PROCEDURE POMELO_BEFORE_DROP_PRIMARY_KEY(IN `SCHEMA_NAME_ARGUMENT` VARCHAR(255), IN `TABLE_NAME_ARGUMENT` VARCHAR(255))
BEGIN
        DECLARE HAS_AUTO_INCREMENT_ID TINYINT(1);
        DECLARE PRIMARY_KEY_COLUMN_NAME VARCHAR(255);
        DECLARE PRIMARY_KEY_TYPE VARCHAR(255);
        DECLARE SQL_EXP VARCHAR(1000);
        SELECT COUNT(*)
                INTO HAS_AUTO_INCREMENT_ID
                FROM `information_schema`.`COLUMNS`
                WHERE `TABLE_SCHEMA` = (SELECT IFNULL(SCHEMA_NAME_ARGUMENT, SCHEMA()))
                        AND `TABLE_NAME` = TABLE_NAME_ARGUMENT
                        AND `Extra` = 'auto_increment'
                        AND `COLUMN_KEY` = 'PRI'
                        LIMIT 1;
        IF HAS_AUTO_INCREMENT_ID THEN
                SELECT `COLUMN_TYPE`
                        INTO PRIMARY_KEY_TYPE
                        FROM `information_schema`.`COLUMNS`
                        WHERE `TABLE_SCHEMA` = (SELECT IFNULL(SCHEMA_NAME_ARGUMENT, SCHEMA()))
                                AND `TABLE_NAME` = TABLE_NAME_ARGUMENT
                                AND `COLUMN_KEY` = 'PRI'
                        LIMIT 1;
                SELECT `COLUMN_NAME`
                        INTO PRIMARY_KEY_COLUMN_NAME
                        FROM `information_schema`.`COLUMNS`
                        WHERE `TABLE_SCHEMA` = (SELECT IFNULL(SCHEMA_NAME_ARGUMENT, SCHEMA()))
                                AND `TABLE_NAME` = TABLE_NAME_ARGUMENT
                                AND `COLUMN_KEY` = 'PRI'
                        LIMIT 1;
                SET SQL_EXP = CONCAT('ALTER TABLE `', (SELECT IFNULL(SCHEMA_NAME_ARGUMENT, SCHEMA())), '`.`', TABLE_NAME_ARGUMENT, '` MODIFY COLUMN `', PRIMARY_KEY_COLUMN_NAME, '` ', PRIMARY_KEY_TYPE, ' NOT NULL;');
                SET @SQL_EXP = SQL_EXP;
                PREPARE SQL_EXP_EXECUTE FROM @SQL_EXP;
                EXECUTE SQL_EXP_EXECUTE;
                DEALLOCATE PREPARE SQL_EXP_EXECUTE;
        END IF;
END;
CALL POMELO_BEFORE_DROP_PRIMARY_KEY(NULL, 'MyTable');
DROP PROCEDURE IF EXISTS POMELO_BEFORE_DROP_PRIMARY_KEY;
ALTER TABLE `MyTable` DROP PRIMARY KEY;
[15:20:31 ERR] [Microsoft.EntityFrameworkCore.Database.Command] Failed executing DbCommand (63,469ms) [Parameters=[], CommandType='Text', CommandTimeout='30']
DROP PROCEDURE IF EXISTS POMELO_BEFORE_DROP_PRIMARY_KEY;
CREATE PROCEDURE POMELO_BEFORE_DROP_PRIMARY_KEY(IN `SCHEMA_NAME_ARGUMENT` VARCHAR(255), IN `TABLE_NAME_ARGUMENT` VARCHAR(255))
BEGIN
        DECLARE HAS_AUTO_INCREMENT_ID TINYINT(1);
        DECLARE PRIMARY_KEY_COLUMN_NAME VARCHAR(255);
        DECLARE PRIMARY_KEY_TYPE VARCHAR(255);
        DECLARE SQL_EXP VARCHAR(1000);
        SELECT COUNT(*)
                INTO HAS_AUTO_INCREMENT_ID
                FROM `information_schema`.`COLUMNS`
                WHERE `TABLE_SCHEMA` = (SELECT IFNULL(SCHEMA_NAME_ARGUMENT, SCHEMA()))
                        AND `TABLE_NAME` = TABLE_NAME_ARGUMENT
                        AND `Extra` = 'auto_increment'
                        AND `COLUMN_KEY` = 'PRI'
                        LIMIT 1;
        IF HAS_AUTO_INCREMENT_ID THEN
                SELECT `COLUMN_TYPE`
                        INTO PRIMARY_KEY_TYPE
                        FROM `information_schema`.`COLUMNS`
                        WHERE `TABLE_SCHEMA` = (SELECT IFNULL(SCHEMA_NAME_ARGUMENT, SCHEMA()))
                                AND `TABLE_NAME` = TABLE_NAME_ARGUMENT
                                AND `COLUMN_KEY` = 'PRI'
                        LIMIT 1;
                SELECT `COLUMN_NAME`
                        INTO PRIMARY_KEY_COLUMN_NAME
                        FROM `information_schema`.`COLUMNS`
                        WHERE `TABLE_SCHEMA` = (SELECT IFNULL(SCHEMA_NAME_ARGUMENT, SCHEMA()))
                                AND `TABLE_NAME` = TABLE_NAME_ARGUMENT
                                AND `COLUMN_KEY` = 'PRI'
                        LIMIT 1;
                SET SQL_EXP = CONCAT('ALTER TABLE `', (SELECT IFNULL(SCHEMA_NAME_ARGUMENT, SCHEMA())), '`.`', TABLE_NAME_ARGUMENT, '` MODIFY COLUMN `', PRIMARY_KEY_COLUMN_NAME, '` ', PRIMARY_KEY_TYPE, ' NOT NULL;');
                SET @SQL_EXP = SQL_EXP;
                PREPARE SQL_EXP_EXECUTE FROM @SQL_EXP;
                EXECUTE SQL_EXP_EXECUTE;
                DEALLOCATE PREPARE SQL_EXP_EXECUTE;
        END IF;
END;
CALL POMELO_BEFORE_DROP_PRIMARY_KEY(NULL, 'MyTable');
DROP PROCEDURE IF EXISTS POMELO_BEFORE_DROP_PRIMARY_KEY;
ALTER TABLE `MyTable` DROP PRIMARY KEY;
```

**Comment by n-gao** [comment]
I'm having the same issue. Is there any plan on fixing this?
Thanks!

**Comment by lauxjpn** [comment]
I will take a closer look at this over the week.

**Comment by lauxjpn** [comment]
This is a nasty one and will likely need to be fixed with yet another stored procedure, due to limited model access in the migration generator.

As was pointed out above, the underlying issue is, that MySQL might *reuse* an existing index when a foreign key is added:

From [13.1.20.6 FOREIGN KEY Constraints: Conditions and Restrictions](https://dev.mysql.com/doc/refman/8.0/en/create-table-foreign-keys.html):
> MySQL requires indexes on foreign keys and referenced keys so that foreign key checks can be fast and not require a table scan. In the referencing table, there must be an index where the foreign key columns are listed as the first columns in the same order. Such an index is created on the referencing table automatically if it does not exist. This index might be silently dropped later if you create another index that can be used to enforce the foreign key constraint.

Also as pointed out above, because of the described behavior, we would need to drop all foreign keys that *might* reuse the index before removing it and then add the foreign keys again.

Because the migration code that generates the drop statement cannot retrieve the columns of the key anymore (because they are not passed to the method that handles the drop and the key itself is not part of the target model anymore) and we cannot access the source model, we cannot retrieve all the foreign keys that need to be temporarily dropped. (Maybe @ajcvickers or @roji can confirm that that is the case.)

So the only option left to us seems to be to add a stored procedure, that enumerates the foreign keys, drops them, drops the actual index that should be dropped and finally adds the foreign keys again.

**Comment by ajcvickers** [comment]
@bricelam is the Migrations guru. If I'm understanding it correctly, then I think we have recently seen a couple of cases similar to this.

**Comment by roji** [comment]
@lauxjpn note that the methods in MigrationsSqlGenerator do accept an IModel parameter, which can be used to check stuff on the model. For example, take a look at [FindEntityTypes](https://github.com/dotnet/efcore/blob/b230bc07c0561c07171c32f71eebb01ee1376fbc/src/EFCore.Relational/Migrations/MigrationsSqlGenerator.cs#L1581) and FindProperty which allow the migrations generator to go back to the model and perform various checks. I'm not sure if that would fulfil your needs here (or if there are any specific pitfalls), as @ajcvickers wrote @bricelam is the guy for migrations.

**Comment by bricelam** [comment]
SQL Server rebuilds indexes during AlterColumnOperation using the backing model:

https://github.com/dotnet/efcore/blob/8401da5d57d4bd769cf349f31e7df3334c825240/src/EFCore.SqlServer/Migrations/SqlServerMigrationsSqlGenerator.cs#L251-L257

And rebuilds default constraints using SQL:

https://github.com/dotnet/efcore/blob/master/src/EFCore.SqlServer/Migrations/SqlServerMigrationsSqlGenerator.cs#L1578-L1606

There are cases where we'll miss dropping an index or miss re-creating a default constraint if it's not part of your model, but if your model is the source of truth (as it should be if you're using Migrations) everything generally works.

Table (and column) rebuilds (https://github.com/dotnet/efcore/issues/329) will also rely heavily on the EF model being accurate and complete.

**Comment by lauxjpn** [comment]
Thanks for the input an samples!

The problem in this case here is, that we need to drop and add some foreign keys, to drop an alternate key (so during a `DropUniqueConstraintOperation`).

To determine the foreign keys to drop and create, we need to know the column names of the alternate key.

The target model parameter does not contain the alternate key anymore, that should be dropped (which makes sense, as it is not part of the target model/snapshot after applying the migration), there is also no way to access the source model (that would still contain the alternate key), and the column names of the alternate key are not part of the `Generate(DropUniqueConstraintOperation, ...)` method parameters. 

I suppose we could just drop and create *all* foreign keys of the table in question instead of just the ones actually necessary.

But I think there would still exists another issue, where if one of the foreign keys gets dropped as part of the same migration (in a later operation), than my understanding of the target model is, that this foreign key would also be already missing from the target model when dropping the alternate key.
This would result in this foreign key not being dropped and created during the `DropUniqueConstraintOperation` which would result in MySQL throwing an error, when trying to drop the alternate key (though it would probably work in all other cases).

**Comment by bricelam** [comment]
This may be a case where it's best to tell users they need to manually edit the migration and add the drop statements themselves. It isn't a goal of EF to hide *all* the details of making schema changes--just to make it considerably easier.

Two additional thoughts:
* We could provide a way to access the previous model in addition to the current one. I'm against this since it's an anti-pattern that would encourage diffing at SQL-generation time.
* This is a good case where the provider could use a better hook (aside from just transforming annotations) during model diffing to produce better migrations.

**Comment by lauxjpn** [comment]
> This is a good case where the provider could use a better hook (aside from just transforming annotations) during model diffing to produce better migrations.

I think that would be a good idea. Being able to influence the actual migration generation process would be helpful for many scenarios.

> This may be a case where it's best to tell users they need to manually edit the migration and add the drop statements themselves.

I think that is acceptable. Dropping alternate or primary keys is part of a group of operations that should be used rarely anyway.

**Comment by lauxjpn** [comment]
Do to the previously anticipated issues with changes to primary keys/unique constraints and foreign keys in the same migrations, we altered the previous fix (#1020) in #1282 (for 5.0.0) to not automatically drop and later recreate all foreign keys.

Instead, we have now introduced an additional parameter for those operations, than can optionally be set in the `Up()` and `Down()` scripts, when calling `DropPrimaryKey()` or `DropUniqueConstraint()`.

In the long run, the preferred solution will be to recreate foreign keys automatically using SQL, but I currently don't have the time to implement it. Once we have implemented it, we will remove the additional parameter again and adding/removing foreign keys will automatically work in all cases.
